docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,303 @@
+passing the dependency results using task objects. This technique of handling
+dependencies make it possible for the ``DependencyBear`` to support caching.
+The ``dependency_results`` lie inside ``self.dependecy_results`` inside a bear
+and can be accessed that way (although this is discouraged since it bypasses
+caching and could yield weird results when the core is run multiple times in a
+row).
+
+This bear serves as a base class which parallelizes tasks for each dependency
+result. A bear dependent on other bears can specify its dependencies in
+``BEAR_DEPS``. For example, there are two bears ``Foo`` and ``Bar``
+and bear ``Bar`` depends on ``Foo``. This can be written as
+
+::
+
+    class BarBear(DependencyBear):
+        BEAR_DEPS = {FooBear}
+
+This solves the dependency issues of GlobalBears on LocalBears that were there
+in the old core. Now that the new dependency management is in place GlobalBears
+won't be stalled due to the termination of a LocalBear run. This eradicates all
+the synchronization problems faced by the old core.
+
+Multiple bears can be included as a dependency of a bear in the BEAR_DEPS
+field. The results of the dependency bears are saved in a dictionary
+called ``_dependency_results`` which is initialized in the ``__init()__``
+method of the class ``Bear.py`` and can be accessed using the method
+``dependency_results()`` also belonging to the same class.
+
+Writing a DependencyBear
+------------------------
+
+Let's consider a bear to be dependent on a project bear ``Fizz`` and a file bear
+``Buzz`` then the corresponding DependencyBear let's call it ``FizzBuzz`` will
+look like the following:
+
+::
+
+    class FizzBuzzBear(DependencyBear):
+        BEAR_DEPS = {FizzBear, BuzzBear}
+
+        def analyze(self, dependency_bear, dependency_result, a_number=100):
+            yield '{} ({}) - {}'.format(
+                dependency_bear.name, a_number, dependency_result)
+
+Ability to modify bear dependencies at runtime
+----------------------------------------------
+
+A bear might depend on multiple bears (to run) before its execution can begin.
+``Bear.BEAR_DEPS`` is just a collection (a set) of classes (Bear classes) that
+need to be executed before that bear can run. Once all these dependencies
+(Bears) have run their results are passed into ``self.dependency_results``.
+The results are in the form of a dictionary with the types of the bears and
+their corresponding results (in the form of a list) as *Key-Value* pairs.
+From the previous example if we try to access the BEAR_DEPS of the ``BarBear``
+we will get the result ``{<class 'coalib.core.Bear.FooBear'>}``.
+
+In the `__init__()` method of the class ``Bear.py`` the dependencies specified
+in the ``BEAR_DEPS`` are copied to every instance of a Bear run using
+``self.BEAR_DEPS = set(self.BEAR_DEPS)`` which makes runtime modifications
+possible.
+
+Override bears
+--------------
+
+A NextGen bear has to have the following functions to perform analysis:
```

Comment:
Any bear that inherits from the `FileBear`, `ProjectBear` or `DependencyBear` classes will already have the `generate_tasks()` method defined in the parent class. In that case a new bear will only need to implement the `analyze()` method. IMHO you should mention that somewhere here. ;-)